In [2]:
# initial setup
try:
    # settings colab:
    import google.colab
    
    # si usan colab, deben cambiar el token de esta url
    #! mkdir -p ../data
    # los que usan colab deben modificar el token de esta url:
    #! wget -O ../data/properati_2017.csv https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_students_2020/master/M2/CLASE_07_Pandas_2/Data/resultado-de-encuestas-2017-2018.csv?token=AA4GFHO3EKGWGILBCV4BISC6WR6B4
    
except ModuleNotFoundError:    
    # settings local:
    %run "common/base_setup.py"

Running command `conda list`... ok
pandas=1.0.3 already installed
matplotlib=2.2.2 already installed
bokeh=2.0.0 already installed
seaborn=0.10.0 already installed
ipywidgets=7.5.1 already installed
pytest=5.3.4 already installed
chardet=3.0.4 already installed
plotly=4.6.0 already installed
Running command `conda install --yes chart-studio=1.1.0`... error
    
PackagesNotFoundError: The following packages are not available from current channels:

  - chart-studio=1.1.0

Current channels:

  - https://repo.anaconda.com/pkgs/main/linux-64
  - https://repo.anaconda.com/pkgs/main/noarch
  - https://repo.anaconda.com/pkgs/r/linux-64
  - https://repo.anaconda.com/pkgs/r/noarch

To search for alternate channels that may provide the conda package you're
looking for, navigate to

    https://anaconda.org

and use the search bar at the top of the page.





## Estructura del Dataset

In [74]:
import pandas as pd
import numpy as np
import re

In [285]:
# low_memorybool, default True
# Internally process the file in chunks, resulting in lower memory use while parsing, 
# but possibly mixed type inference. To ensure no mixed types either set False, 
# or specify the type with the dtype parameter. 
# Note that the entire file is read into a single DataFrame regardless, 
# use the chunksize or iterator parameter to return the data in chunks. (Only valid with C parser).
data = pd.read_csv("./data/properati_2017.csv", sep = ",", low_memory=False)

In [286]:
data.shape

(121220, 26)

In [287]:
data.columns

Index(['Unnamed: 0', 'operation', 'property_type', 'place_name',
       'place_with_parent_names', 'country_name', 'state_name', 'geonames_id',
       'lat-lon', 'lat', 'lon', 'price', 'currency',
       'price_aprox_local_currency', 'price_aprox_usd', 'surface_total_in_m2',
       'surface_covered_in_m2', 'price_usd_per_m2', 'price_per_m2', 'floor',
       'rooms', 'expenses', 'properati_url', 'description', 'title',
       'image_thumbnail'],
      dtype='object')

In [288]:
data.dtypes

Unnamed: 0                      int64
operation                      object
property_type                  object
place_name                     object
place_with_parent_names        object
country_name                   object
state_name                     object
geonames_id                   float64
lat-lon                        object
lat                           float64
lon                           float64
price                         float64
currency                       object
price_aprox_local_currency    float64
price_aprox_usd               float64
surface_total_in_m2           float64
surface_covered_in_m2         float64
price_usd_per_m2              float64
price_per_m2                  float64
floor                         float64
rooms                         float64
expenses                      float64
properati_url                  object
description                    object
title                          object
image_thumbnail                object
dtype: objec

### Columnas Irrelevantes para el estudio - Fase 1

In [289]:
all_columns = set(data.columns)
relevant_columns = all_columns.difference({'Unnamed: 0', 'image_thumbnail', 'properati_url'})
data = data.loc[:, list(relevant_columns)]

### Visualización Preliminar

In [290]:
data.head(10)

rooms     price        lat                    state_name property_type  \
0    NaN   62000.0 -34.661824               Capital Federal            PH   
1    NaN  150000.0 -34.903883        Bs.As. G.B.A. Zona Sur     apartment   
2    NaN   72000.0 -34.652262               Capital Federal     apartment   
3    NaN   95000.0 -34.647797               Capital Federal            PH   
4    NaN   64000.0 -38.002626  Buenos Aires Costa Atlántica     apartment   
5    NaN       NaN -33.014071                    Entre Ríos         house   
6    NaN  130000.0 -34.532957      Bs.As. G.B.A. Zona Norte            PH   
7    NaN  138000.0 -34.559873               Capital Federal     apartment   
8    NaN  195000.0 -34.559873               Capital Federal     apartment   
9    NaN  460000.0 -32.942031                      Santa Fe         house   

  country_name  floor  price_usd_per_m2  price_aprox_local_currency  \
0    Argentina    NaN       1127.272727                  1093959.00   
1    Argentina    NaN               NaN                  2646675.00   
2    Argentina    NaN       1309.090909                  1270404.00   
3    Argentina    NaN               NaN                  1676227.50   
4    Argentina    NaN       1828.571429                  1129248.00   
5    Argentina    NaN               NaN                         NaN   
6    Argentina    NaN       1226.415094                  2293785.00   
7    Argentina    NaN       3066.666667                  2434941.00   
8    Argentina    NaN       3000.000000                  3440677.50   
9    Argentina    NaN               NaN                   455201.45   

                                         description  ...        lon  \
0  2 AMBIENTES TIPO CASA PLANTA BAJA POR PASILLO,...  ... -58.508839   
1  Venta de departamento en décimo piso al frente...  ... -57.964330   
2  2 AMBIENTES 3ER PISO LATERAL LIVING COMEDOR AM...  ... -58.522982   
3  PH 3 ambientes con patio. Hay 3 deptos en lote...  ... -58.516424   
4  DEPARTAMENTO CON FANTÁSTICA ILUMINACIÓN NATURA...  ... -57.549447   
5  Casa en el perímetro del barrio 338, ubicada e...  ... -58.519828   
6  MUY BUEN PH AL FRENTE CON ENTRADA INDEPENDIENT...  ... -58.521782   
7  EXCELENTE MONOAMBIENTE A ESTRENAR AMPLIO SUPER...  ... -58.443362   
8  EXCELENTE DOS AMBIENTES  ESTRENAR AMPLIO SUPER...  ... -58.443362   
9  MEDNOZA AL 7600A UNA CUADRA DE CALLE MENDOZAWH...  ... -60.725919   

   operation  price_aprox_usd  currency  \
0       sell         62000.00       USD   
1       sell        150000.00       USD   
2       sell         72000.00       USD   
3       sell         95000.00       USD   
4       sell         64000.00       USD   
5       sell              NaN       NaN   
6       sell        130000.00       USD   
7       sell        138000.00       USD   
8       sell        195000.00       USD   
9       sell         25798.49       ARS   

                                               title  expenses  \
0                 2 AMB TIPO CASA SIN EXPENSAS EN PB       NaN   
1  VENTA Depto 2 dorm. a estrenar 7 e/ 36 y 37   ...       NaN   
2           2 AMB 3ER PISO CON ASCENSOR APTO CREDITO       NaN   
3                       PH 3 amb. cfte. reciclado          NaN   
4  DEPTO 2 AMB AL CONTRAFRENTE ZONA CENTRO/PLAZA ...       NaN   
5  Casa Barrio 338. Sobre calle 3 de caballería, ...       NaN   
6  MUY BUEN PH AL FRENTE DOS DORMITORIOS , PATIO,...       NaN   
7  JOSE HERNANDEZ 1400 MONOAMBIENTE  ESTRENAR CAT...       NaN   
8  JOSE HERNANDEZ   1400 DOS AMBIENTES ESTRENAR ,...       NaN   
9         WHITE 7637 - 2 DORMITORIOS CON PATIO             NaN   

                   lat-lon    place_name  price_per_m2  \
0  -34.6618237,-58.5088387     Mataderos   1550.000000   
1  -34.9038831,-57.9643295      La Plata           NaN   
2  -34.6522615,-58.5229825     Mataderos   1309.090909   
3  -34.6477969,-58.5164244       Liniers           NaN   
4  -38.0026256,-57.5494468        Centro   1828.571429   
5   -33.0140714,-58.519828  Gua

### Distribución de N/A

In [291]:
def show_null_distribution(df):
    nulls = df.isnull().sum(axis=0)
    nulls_percent = nulls / df.index.size
    nulls_df = pd.concat([nulls, nulls_percent], axis=1)
    nulls_df.columns = ['total', 'percent']
    return nulls_df[nulls_df['total'] > 0] # solo me interesan las columnas que tengan N/A

show_null_distribution(data)

total   percent
rooms                        73830  0.609058
price                        20410  0.168372
lat                          51550  0.425260
floor                       113321  0.934837
price_usd_per_m2             52603  0.433947
price_aprox_local_currency   20410  0.168372
description                      2  0.000016
surface_total_in_m2          39328  0.324435
geonames_id                  18717  0.154405
surface_covered_in_m2        19907  0.164222
lon                          51550  0.425260
price_aprox_usd              20410  0.168372
currency                     20411  0.168380
expenses                    106958  0.882346
lat-lon                      51550  0.425260
place_name                      23  0.000190
price_per_m2                 33562  0.276869

## Tratamiento de Valores Faltantes

### Columna *place_name*

Observamos que la columna *place_with_parent_names* se construye concatenando *country_name*, *state_name* y *place_name*, entonces chequeamos esa columna contra los faltantes:

In [292]:
na_place_name_mask = data["place_name"].isnull()
na_place_name      = data[na_place_name_mask][["place_name", "place_with_parent_names"]]
na_place_name.drop_duplicates()

place_name                      place_with_parent_names
6489        NaN  |Argentina|Bs.As. G.B.A. Zona Norte|Tigre||

Vemos que en este caso los N/A son de **Tigre** 

In [293]:
data[data["place_name"].isnull()] = "Tigre"

### Columnas *description* y *title*

In [294]:
na_description_mask = data["description"].isnull()
na_description      = data[na_description_mask]
na_description

rooms     price      lat                state_name property_type  \
10664   NaN  2.35e+06 -31.4208                   Córdoba     apartment   
96691   NaN     98000      NaN  Bs.As. G.B.A. Zona Norte     apartment   

      country_name floor price_usd_per_m2 price_aprox_local_currency  \
10664    Argentina   NaN              NaN                2.32549e+06   
96691    Argentina   NaN              NaN                1.72916e+06   

      description  ...      lon operation price_aprox_usd currency  \
10664         NaN  ... -64.4992      sell          131797      ARS   
96691         NaN  ...      NaN      sell           98000      USD   

                                             title expenses  \
10664  HERMOSO DEPARTAMENTO CON PISCINA Y COCHERA       NaN   
96691                  Departamento - San Fernando      NaN   

                       lat-lon        place_name price_per_m2  \
10664  -31.4207828,-64.4992141  Villa Carlos Paz          NaN   
96691                      NaN      San Fernando      1781.82   

                                 place_with_parent_names  
10664               |Argentina|Córdoba|Villa Carlos Paz|  
96691  |Argentina|Bs.As. G.B.A. Zona Norte|San Fernando|  

[2 rows x 23 columns]

Por ahora, solo las normalizamos:

In [295]:
rex = re.compile(r'\s+')
fformat = lambda x: rex.sub(' ', str(x)).strip().lower()
data['description'] = data['description'].apply(fformat)
data['title']       = data['title'].apply(fformat)

### Columna *expenses*
En este caso, chequeamos la cantidad de N/A contra la cantidad de posibles calculables.
El criterio de la esa posibilidad, es simple: la palabra **expensa** dentro de *title* o *description*.

In [296]:
na_expenses_mask = data['expenses'].isnull()
na_expenses = data[na_expenses_mask]
na_expenses_total = na_expenses_mask.sum()
calc_expenses_total = na_expenses['title'].str.contains('expensa', case=False, regex=False).sum() + \
   na_expenses['description'].str.contains('expensa', case=False, regex=False).sum() 

print("Proportion of missing expenses: {0}".format(na_expenses_total / data.index.size))
print("Proportion of posibles calculable expenses: {0}".format(calc_expenses_total / na_expenses_total))

Proportion of missing expenses: 0.8821564098333609
Proportion of posibles calculable expenses: 0.12411277879085426


De lo anterior, concluimos que casi un 90% de los datos son N/A y de ese porcentage menos de un 13% tiene alguna posibilidad de ser inferido. Por ende, no la vamos a tener en cuenta para hacer el estudio.

### Columna *floor*
Al igual que en el caso anterior, lo primero es chequear la proporción de N/A.

In [297]:
na_floor_mask = data['floor'].isnull()
na_floor = data[na_floor_mask]
na_floor_total = na_floor.index.size
print("Proportion of missing floors: {0}".format(na_floor_total / data.index.size))

Proportion of missing floors: 0.9346477478963867


Es una proporción alta aunque creemos que este campo tiene más relevancia cuando el *property_type* es **apartment**.

In [298]:
data_apartment_df = data[data['property_type'] == 'apartment']
na_apartment_floor_mask = data_apartment_df['floor'].isnull()
na_apartment_floor_total = na_apartment_floor_mask.sum()
print("Total of apartment properties: {0}".format(data_apartment_df.index.size))
print("Proportion of missing floors: {0}".format(na_apartment_floor_total / data_apartment_df.index.size))

Total of apartment properties: 71064
Proportion of missing floors: 0.9082376449397725


Nuevamente, la proporción de N/A es muy alta. Por ende, esta columna también la vamos a descartar del estudio.

### Columna *rooms*
De nuevo, chequeamos la proporción de N/A.

In [299]:
na_rooms_mask = data['rooms'].isnull()
na_rooms = data[na_rooms_mask]
na_rooms_total = na_rooms.index.size
print("Proportion of missing rooms: {0}".format(na_rooms_total / data.index.size))

Proportion of missing rooms: 0.6089341692789969


Otra vez, nos encontramos con una proporción alta (aunque no tal alta como los casos anteriores).
En este caso, la descripción en muchos casos tiene información que podría ser usada para inferir la cantidad de ambientes que tiene una propiedad pero habría que programar lógica capaz de extraer dicha información, sumarizarla y en muchos casos chequearla personalmente.

Desestimamos incluir esta columna en este primer análisis debido a la carga de trabajo que supone lo explicado en el párrafo anterior.

### Columna *price* y relacionadas
Vamos a dividir el dataset en 2 partes:
 * el precio está disponible o puede inferirse a partir de otras columnas.
 * el precio es N/A y las columnas de las que se podría inferir también son N/A.

In [300]:
def show_prices_na(df):
    price_na_mask   = df[['price', 'price_aprox_local_currency', 'price_aprox_usd']].isna().all(axis=1)
    surface_na_mask = df[['surface_total_in_m2', 'surface_covered_in_m2']].isna().all(axis=1)
    pper_m2_na_mask = df[['price_per_m2', 'price_usd_per_m2']].isna().all(axis=1)
    uncalculable_price_mask = price_na_mask & (surface_na_mask | pper_m2_na_mask)
    df_uncalculables = df[uncalculable_price_mask]
    df_calculables = df[~uncalculable_price_mask]
    print("Missing prices (uncalculable): {0}".format(df_uncalculables.index.size))
    print("Populated or calculable prices: {0}".format(df_calculables.index.size))
show_prices_na(data)

Missing prices (uncalculable): 20410
Populated or calculable prices: 100810


Lo anterior significa que tenemos 20410 registros que carecen de los datos necesarios para ser tenidos en cuenta. Vamos a tratar de inferir algunos usando las columnas **title** y **description**.

In [301]:
def apply_price_from_title(r):
    if re.search(r'en\s+venta', r.title, re.I):
        if r.title.startswith(r'u$d') or r.title.startswith(r'usd'):
            r.price    = np.float64(r.title.split()[1].strip().replace('.', ''))
            r.currency = 'USD'
        elif r.title.startswith(r'$'):
            r.price    = np.float64(r.title.split()[1].strip().replace('.', ''))
            r.currency = 'ARS'
    return r

price_na_mask   = data[['price', 'price_aprox_local_currency', 'price_aprox_usd']].isna().all(axis=1)
surface_na_mask = data[['surface_total_in_m2', 'surface_covered_in_m2']].isna().all(axis=1)
pper_m2_na_mask = data[['price_per_m2', 'price_usd_per_m2']].isna().all(axis=1)
uncalculable_price_mask = price_na_mask & (surface_na_mask | pper_m2_na_mask)

data[uncalculable_price_mask] = data[uncalculable_price_mask].apply(apply_price_from_title, axis=1)
show_prices_na(data)

Missing prices (uncalculable): 17792
Populated or calculable prices: 103428


### Columnas Irrelevantes para el estudio - Fase 2
Procedemos a quitar del dataset las columnas que *expenses* y *floor*.
Quitamos además las columnas *operation* y *country_name* (tienen un solo valor) y *place_with_parent_names* dado que es la concatenación de otras 3 columnas.

In [302]:
all_columns = set(data.columns)
relevant_columns = all_columns.difference({'country_name', 'operation', 'place_with_parent_names', 'expenses', 'floor'})
data = data.loc[:, list(relevant_columns)]
data.head()

rooms      lat surface_total_in_m2  geonames_id surface_covered_in_m2  \
0   NaN -34.6618                  55  3.43079e+06                    40   
1   NaN -34.9039                 NaN  3.43204e+06                   NaN   
2   NaN -34.6523                  55  3.43079e+06                    55   
3   NaN -34.6478                 NaN  3.43133e+06                   NaN   
4   NaN -38.0026                  35  3.43555e+06                    35   

       lon price_aprox_usd currency                    state_name  \
0 -58.5088           62000      USD               Capital Federal   
1 -57.9643          150000      USD        Bs.As. G.B.A. Zona Sur   
2  -58.523           72000      USD               Capital Federal   
3 -58.5164           95000      USD               Capital Federal   
4 -57.5494           64000      USD  Buenos Aires Costa Atlántica   

  price_aprox_local_currency  \
0                1.09396e+06   
1                2.64668e+06   
2                 1.2704e+06   
3                1.67623e+06   
4                1.12925e+06   

                                               title property_type  \
0                 2 amb tipo casa sin expensas en pb            PH   
1        venta depto 2 dorm. a estrenar 7 e/ 36 y 37     apartment   
2           2 amb 3er piso con ascensor apto credito     apartment   
3                          ph 3 amb. cfte. reciclado            PH   
4  depto 2 amb al contrafrente zona centro/plaza ...     apartment   

  price_usd_per_m2   price                  lat-lon place_name price_per_m2  \
0          1127.27   62000  -34.6618237,-58.5088387  Mataderos         1550   
1              NaN  150000  -34.9038831,-57.9643295   La Plata          NaN   
2          1309.09   72000  -34.6522615,-58.5229825  Mataderos      1309.09   
3              NaN   95000  -34.6477969,-58.5164244    Liniers          NaN   
4          1828.57   64000  -38.0026256,-57.5494468     Centro      1828.57   

                                         description  
0  2 ambientes tipo casa planta baja por pasillo,...  
1  venta de departamento en décimo piso al frente...  
2  2 ambientes 3er piso lateral living comedor am...  
3  ph 3 ambientes con patio. hay 3 deptos en lote...  
4  departamento con fantástica iluminación natura...

## Tratamiento de duplicados
El primer lote de duplicados que consideramos se compone por el siguiente conjunto de columnas:
 * property_type
 * lat
 * lon
 * price
 * title

In [303]:
data = data.drop_duplicates(subset=('property_type', 'lat', 'lon', 'geonames_id', 'price', 'place_name'), keep='first')
data.shape

(86646, 18)

In [284]:
show_null_distribution(data)

total   percent
rooms                       45024  0.523791
lat                         23413  0.272377
surface_total_in_m2         28098  0.326881
surface_covered_in_m2       13726  0.159683
lon                         23413  0.272377
price_aprox_usd              9379  0.109111
currency                     7171  0.083424
price_aprox_local_currency   9379  0.109111
price_usd_per_m2            33365  0.388155
price                        7170  0.083413
lat-lon                     23413  0.272377
price_per_m2                19208  0.223458